### PR TITLE
Expand ArmorStand API

### PR DESCRIPTION
**Pull request by willies952002**
Add the following:
- Add proper methods for getting and setting items in both hands. Deprecates old methods
- Enable/Disable slot interactions

### CODE REVIEW

File: Spigot-API-Patches/0127-Expand-ArmorStand-API.patch [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+@@ -10,6 +13,10 @@ public interface ArmorStand extends LivingEntity {
+      * currently holding
+      *
+      * @return the held item
++     // Paper start - Deprecate in favor of setItemInMainHand
++     * @deprecated use {@link ArmorStand#getItemInMainHand()} instead
++     * @see ArmorStand#getItemInMainHand()
++     // Paper end
+      */
+     ItemStack getItemInHand();
```
Comment (kashike) [inline Spigot-API-Patches/0127-Expand-ArmorStand-API.patch]: Please add the `@Deprecated` annotation.

File: Spigot-API-Patches/0127-Expand-ArmorStand-API.patch [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+@@ -18,6 +25,10 @@ public interface ArmorStand extends LivingEntity {
+      * holding
+      *
+      * @param item the item to hold
++     // Paper start - Deprecate in favor of setItemInMainHand
++     * @deprecated use {@link ArmorStand#setItemInMainHand(ItemStack)} instead
++     * @see ArmorStand#setItemInMainHand(ItemStack)
++     // Paper end
+      */
+     void setItemInHand(ItemStack item);
```
Comment (kashike) [inline Spigot-API-Patches/0127-Expand-ArmorStand-API.patch]: Please add the `@Deprecated` annotation.

File: Spigot-API-Patches/0127-Expand-ArmorStand-API.patch [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+      */
+     void setCanMove(boolean move);
++
++    /**
++     * Returns the item the armor stand is
++     * currently holding in it's main hand
++     *
++     * @return the held item
++     */
++    ItemStack getItemInMainHand();
```
Comment (kashike) [inline Spigot-API-Patches/0127-Expand-ArmorStand-API.patch]: I'd like to see a method like https://github.com/PaperMC/Paper/blob/master/Spigot-Server-Patches/0091-Access-items-by-EquipmentSlot.patch#L18-L24:
```java
ItemStack getItemInHand(final org.bukkit.inventory.EquipmentSlot slot);
```

File: Spigot-API-Patches/0127-Expand-ArmorStand-API.patch [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
++
++    /**
++     * Sets the item the armor stand is currently
++     * holding in it's main hand
++     *
++     * @param item the item to hold
++     */
++    void setItemInMainHand(ItemStack item);
```
Comment (kashike) [inline Spigot-API-Patches/0127-Expand-ArmorStand-API.patch]: I'd like to see a method like https://github.com/PaperMC/Paper/blob/master/Spigot-Server-Patches/0091-Access-items-by-EquipmentSlot.patch#L39-L47:
```java
ItemStack setItemInHand(final org.bukkit.inventory.EquipmentSlot slot, final ItemStack item);
```

File: Spigot-API-Patches/0127-Expand-ArmorStand-API.patch [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
++     * @return integer
++     */
++    List<EquipmentSlot> getDisabledSlots();
++
++    /**
++     * Set the disabled slots
++     *
++     * @param slots var-arg array of slots to lock
++     */
++    void setDisabledSlots(EquipmentSlot... slots);
```
Comment (kashike) [inline Spigot-API-Patches/0127-Expand-ArmorStand-API.patch]:
```java
void addDisabledSlot(final EquipmentSlot... slots);

void removedDisabledSlot(final EquipmentSlot... slots);
```

**Review by kashike — CHANGES_REQUESTED** [review]

File: Spigot-API-Patches/0127-Expand-ArmorStand-API.patch [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
++     * @param item the item to hold
++     */
++    void setItemInHand(final EquipmentSlot slot, final ItemStack item);
++
++    /**
++     * Get the list of disabled slots
++     *
++     * @return integer
++     */
++    List<EquipmentSlot> getDisabledSlots();
```
Comment (kashike) [inline Spigot-API-Patches/0127-Expand-ArmorStand-API.patch]: Should be a `Set`.

**Review by kashike — CHANGES_REQUESTED** [review]

File: Spigot-API-Patches/0127-Expand-ArmorStand-API.patch (hunk 6 above)
Comment (willies952002) [reply Spigot-API-Patches/0127-Expand-ArmorStand-API.patch]: in what way..?

File: Spigot-API-Patches/0127-Expand-ArmorStand-API.patch [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+- Enable/Disable slot interactions
+
+diff --git a/src/main/java/org/bukkit/entity/ArmorStand.java b/src/main/java/org/bukkit/entity/ArmorStand.java
+index 099da6ce..c669b490 100644
+--- a/src/main/java/org/bukkit/entity/ArmorStand.java
++++ b/src/main/java/org/bukkit/entity/ArmorStand.java
+@@ -1,8 +1,11 @@
+ package org.bukkit.entity;
+ 
++import org.bukkit.inventory.EquipmentSlot; // Paper
```
Comment (kashike) [inline Spigot-API-Patches/0127-Expand-ArmorStand-API.patch]: Avoid the import statement - fully qualify below.

File: Spigot-API-Patches/0127-Expand-ArmorStand-API.patch [hunk 8]
```diff
@@ -0,0 +1,146 @@
...
+ import org.bukkit.inventory.ItemStack;
+ import org.bukkit.util.EulerAngle;
+ 
++import java.util.Set; // Paper
```
Comment (kashike) [inline Spigot-API-Patches/0127-Expand-ArmorStand-API.patch]: Avoid the import statement - fully qualify below.

File: Spigot-Server-Patches/0315-Implement-Expanded-ArmorStand-API.patch [hunk 9]
```diff
@@ -0,0 +1,144 @@
...
++    public void addDisabledSlots(org.bukkit.inventory.EquipmentSlot... slots) {
++        java.util.Set<org.bukkit.inventory.EquipmentSlot> disabled = getDisabledSlots();
++        disabled.addAll(java.util.Arrays.asList(slots));
++        setDisabledSlots(disabled.toArray(new org.bukkit.inventory.EquipmentSlot[0]));
++    }
++
++    @Override
++    public void removeDisabledSlots(org.bukkit.inventory.EquipmentSlot... slots) {
++        java.util.Set<org.bukkit.inventory.EquipmentSlot> disabled = getDisabledSlots();
++        disabled.removeAll(java.util.Arrays.asList(slots));
```
Comment (kashike) [inline Spigot-Server-Patches/0315-Implement-Expanded-ArmorStand-API.patch]: Just do a normal for loop here:
```java
for (final org.bukkit.inventory.EquipmentSlot slot : slots) disabled.remove(slot);
```

**Review by kashike — CHANGES_REQUESTED** [review]

File: Spigot-Server-Patches/0315-Implement-Expanded-ArmorStand-API.patch [hunk 10]
```diff
@@ -0,0 +1,144 @@
...
++            net.minecraft.server.EnumItemSlot nmsSlot = org.bukkit.craftbukkit.CraftEquipmentSlot.getNMS(slot);
++            disabled += (1 << nmsSlot.c()) + (1 << (nmsSlot.c() + 8)) + (1 << (nmsSlot.c() + 16));
++        }
++        getHandle().setDisabledSlots(disabled);
++    }
++
++    @Override
++    public void addDisabledSlots(org.bukkit.inventory.EquipmentSlot... slots) {
++        java.util.Set<org.bukkit.inventory.EquipmentSlot> disabled = getDisabledSlots();
++        disabled.addAll(java.util.Arrays.asList(slots));
```
Comment (kashike) [inline Spigot-Server-Patches/0315-Implement-Expanded-ArmorStand-API.patch]: `java.util.Collections.addAll(disabled, slots);`

**Review by kashike — DISMISSED** [review]

File: Spigot-Server-Patches/0315-Implement-Expanded-ArmorStand-API.patch [hunk 11]
```diff
@@ -0,0 +1,144 @@
...
++        return getEquipment().getItemInOffHand();
++    }
++
++    @Override
++    public void setItemInOffHand(ItemStack item) {
++        getEquipment().setItemInOffHand(item);
++    }
++
++    @Override
++    public ItemStack getItemInHand(org.bukkit.inventory.EquipmentSlot slot) {
```
Comment (aikar) [inline Spigot-Server-Patches/0315-Implement-Expanded-ArmorStand-API.patch]: Actually, we should very well make this setItem/getItem(slot) and make all the slots work.

**Review by aikar — CHANGES_REQUESTED** [review]

**Comment by willies952002** [comment]
is there any other changes that need to be made?

File: Spigot-API-Patches/0127-Expand-ArmorStand-API.patch [hunk 12]
```diff
@@ -0,0 +1,134 @@
...
++    /**
++     * Sets the item the armor stand has
++     * equip in the given equipment slot
++     *
++     * @param slot the equipment slot to set
++     * @param item the item to hold
++     */
++    void setItem(final org.bukkit.inventory.EquipmentSlot slot, final ItemStack item);
++
++    /**
```
Comment (aikar) [inline Spigot-API-Patches/0127-Expand-ArmorStand-API.patch]: I think all of the disabled slots API's should describe just what this API does so people don't have to go dig into technical wiki articles on what disabled slots do.

I don't even know!

File: Spigot-API-Patches/0127-Expand-ArmorStand-API.patch (hunk 3 above)
Comment (aikar) [reply Spigot-API-Patches/0127-Expand-ArmorStand-API.patch]: I don't think we need these API's anymore with .getItem and setItem, it introduces confusion on "are these different", so let's just remove the hand specific ones.

Update the deprecated tags above too.

After that, I think we're good to merge.

**Review by aikar — CHANGES_REQUESTED** [review]

File: Spigot-API-Patches/0127-Expand-ArmorStand-API.patch (hunk 12 above)
Comment (willies952002) [reply Spigot-API-Patches/0127-Expand-ArmorStand-API.patch]: `This makes it so a player is unable to place, remove, or replace an item in the given slot(s)`

does this seem like it's not enough of a description?
For some reason I feel this needs a slightly longer explanation

Comment (aikar) [reply Spigot-API-Patches/0127-Expand-ArmorStand-API.patch]: maybe something like "This makes it so a player is unable to interact with the Armor Stand to place, remove, or replace an item in the given slot(s)."

I assume the items fall if the armor stand is broken still, so may be worth noting that expectation, that once a slot is disabled, item retrieval can only be done on break?

**Review by aikar — APPROVED** [review]

**Pull request merged**